src/Stratis.Bitcoin.Features.SmartContracts/PoW/SmartContractBlockDefinition.cs

```diff
@@ -86,12 +89,28 @@ public override void AddToBlock(TxMempoolEntry mempoolEntry)
             {
                 this.logger.LogDebug("Transaction contains smart contract information.");
 
-                if (this.blockGasConsumed >= GasPerBlockLimit) 
+                if (this.blockGasConsumed >= GasPerBlockLimit)
+                {
+                    this.logger.LogDebug("The gas limit for this block has been reached.");
                     return;
+                }
 
-                // We HAVE to firstly execute the smart contract contained in the transaction
-                // to ensure its validity before we can add it to the block.
                 IContractExecutionResult result = this.ExecuteSmartContract(mempoolEntry);
+
+                // If including this transaction would put us over the block gas limit, then don't include it
+                // and roll back all of the execution we did.
+                if (this.blockGasConsumed >= GasPerBlockLimit)
```

Comment:
If this puts us at the limit exactly, then it's fine.

```suggestion
                if (this.blockGasConsumed > GasPerBlockLimit)
```